bessctl/static/pipeline.js

```diff
@@ -0,0 +1,177 @@
+    for (var i = 0; i < gates.length; i++) {
+        var gate = gates[i];
+        key = [module_name, gate_type, gate[gate_type]];
+        value = {timestamp: gate.timestamp,
+                 bits: Number(gate.bytes * 8),
+                 pkts: Number(gate.pkts),
+                 cnt: Number(gate.cnt),
+                 batchsize: gate.cnt ? gate.pkts / gate.cnt : 0};
+        if (!(key in stats)) {
+            stats[key] = [value];
+        } else {
+            stats[key].push(value);
+        }
+    }
+}
+
+function get_edge_label(stats) {
+    var num_stats = stats.length;
+    var value = stats[num_stats - 1];
+    var label = '?'
+
+    if (value.timestamp > 0) {
+        switch (opt_mode) {
+            case 'total':
+                label = value[opt_field];
+                break;
+            case 'rate':
+                if (num_stats >= 2) {
+                    last = stats[num_stats - 2];
+                    time_diff = value.timestamp - last.timestamp;
+                    if (opt_field == 'batchsize') {
+                        packets = value.pkts - last.pkts;
+                        batches = value.cnt - last.cnt;
+                        label = batches ? packets / batches : 'N/A';
+                    } else {
+                        value_diff = value[opt_field] - last[opt_field];
+                        label = Math.round(value_diff / time_diff);
+                    }
+                }
+                break;
+            case 'none':
+                return '';
+            default:
+                throw new Error('Unknown mode ' + opt_mode);
+        }
+    }
+
+    if ((typeof label == 'number') && opt_humanreadable) {
+        var unit = ' ';
+        if (opt_mode == 'rate') {
+            if (label > 1000000000) {
+                label /= 1000000000;
+                unit += 'G';
+            } else if (label > 1000000) {
+                label /= 1000000;
+                unit += 'M';
+            } else if (label > 1000) {
+                label /= 1000;
+                unit += 'k';
+            }
+            if (opt_field == 'pkts') {
+                unit += 'pps';
+            } else if (opt_field == 'bits') {
+                unit += 'bps';
+            }
+        }
+        label = label.toLocaleString('en-US', {maximumFractionDigits: 2}) + unit;
+    }
+
+    // We need this HTML hack to give background color to edge labels
+    return `<<table border="0" cellpadding="0"><tr><td bgcolor="white">${label}</td></tr></table>>`;
+}
+
+function graph_to_dot(modules) {
+    opt_field = document.querySelector('input[name="metric"]:checked').value;
+    opt_mode = document.querySelector('input[name="mode"]:checked').value;
+    opt_humanreadable = document.querySelector('input[name="humanreadable"]').checked;
+
+    var nodes = ''
+    for (var module_name in modules) {
+        module = modules[module_name];
+        // add_datapoints(stats, module_name, module.igates, 'igate')
```

Comment:
Do you intentionally comment this line out? or having a plan to add the option or it?

Reply:
Yeah, at the moment it only shows the statistics numbers from output gates. Per-input-gate stats seemed less useful in general, so I didn't bother displaying them... so I didn't want to keep collecting the data. At some point I could add an option to show igate stats, but I have no plan now.